src/goto-programs/remove_exceptions.cpp

```diff
@@ -0,0 +1,598 @@
+        t_null->code=code_assignt(
+          lhs_expr_null,
+          typecast_exprt(rhs_expr_null, lhs_expr_null.type()));
+        t_null->function=instr_it->function;
+
+        // add the assignment exc=f#exception_value
+        symbol_exprt rhs_expr_exc;
+        rhs_expr_exc.set_identifier(id2string(function_id)+EXC_SUFFIX);
+        rhs_expr_exc.type()=function_symbol.type;
+
+        const exprt &lhs_expr_exc=static_cast<const exprt &>(exception);
+        goto_programt::targett t_exc=goto_program.insert_after(instr_it);
+        t_exc->make_assignment();
+        t_exc->source_location=instr_it->source_location;
+        t_exc->code=code_assignt(
+          typecast_exprt(lhs_expr_exc, rhs_expr_exc.type()),
+          rhs_expr_exc);
+        t_exc->function=instr_it->function;
+      }
+
+      instr_it->make_skip();
+    }
+  }
+}
+
+/*******************************************************************\
+
+Function: remove_exceptionst::add_gotos_throw
+
+Inputs:
+
+Outputs:
+
+Purpose: instruments each throw with conditional GOTOS to the 
+         corresponding exception handlers
+
+\*******************************************************************/
+
+void remove_exceptionst::add_throw_gotos(
+  const goto_functionst::function_mapt::iterator &func_it,
+  const goto_programt::instructionst::iterator &instr_it,
+  const remove_exceptionst::stack_catcht &stack_catch)
+{
+  assert(instr_it->type==THROW);
+
+  goto_programt &goto_program=func_it->second.body;
+
+  assert(instr_it->code.operands().size()==1);
+
+  // make sure that these are always forward gotos
+  assert(!instr_it->is_backwards_goto());
+
+  // find the end of the function
+  goto_programt::targett new_state;
+  for(new_state=instr_it; !new_state->is_end_function(); new_state++) {}
+  if(new_state!=instr_it)
+  {
+    // jump to the end of the function
+    // this will appear after the GOTO-based dynamic dispatch below
+    goto_programt::targett t_end=goto_program.insert_after(instr_it);
+    t_end->make_goto(new_state);
+    t_end->source_location=instr_it->source_location;
+    t_end->function=instr_it->function;
+  }
+
+  // add GOTOs implementing the dynamic dispatch of the
+  // exception handlers
+  for(std::size_t i=stack_catch.size(); i-->0;)
+  {
+    for(std::size_t j=stack_catch[i].size(); j-->0;)
+    {
+      goto_programt::targett new_state_pc=
+        stack_catch[i][j].second;
+
+      // find handler
+      goto_programt::targett t_exc=goto_program.insert_after(instr_it);
+      t_exc->make_goto(new_state_pc);
+      t_exc->source_location=instr_it->source_location;
+      t_exc->function=instr_it->function;
+
+      // use instanceof to check that this is the correct handler
+      symbol_typet type(stack_catch[i][j].first);
+      type_exprt expr(type);
+      // find the symbol corresponding to the caught exceptions
+      exprt exc_symbol=instr_it->code;
+      while(exc_symbol.id()!=ID_symbol)
+        exc_symbol=exc_symbol.op0();
```

Comment:
Same question as above (well, also same code -> refactor?): why does one need to iterate here?